src/Microsoft.OData.Core/Uri/ODataUriSlim.cs

```diff
@@ -0,0 +1,139 @@
+﻿// <copyright file="ODataUriSlim.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using Microsoft.OData.UriParser.Aggregation;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A lightweight version of <see cref="ODataUri"/> that is
+    /// used by the writer to efficiently create a copy of the URI when entering
+    /// each scope. This is used to avoid cloning <see cref="ODataUri"/>
+    /// excessively.
+    /// </summary>
+    internal struct ODataUriSlim
+    {
+        private readonly ODataUri odataUri;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUriSlim"/>
+        /// by copying the provided <see cref="ODataUri"/> instance.
+        /// </summary>
+        /// <param name="odataUri">The URI to copy.</param>
+        public ODataUriSlim(ODataUri odataUri)
+        {
+            Debug.Assert(odataUri != null, "odataUri != null");

```

Comment:
Maybe throw an `ArgumentNullException`?

Reply:
I'm a bit conflicted since this add null check for something that is called often but should never be null unless we have an internal bug (not a user-input error).